src/zeroconf/_services/browser.py

```diff
@@ -250,43 +295,83 @@ def _generate_first_next_time(self, now: float_) -> None:
         also delay the first query of the series by a randomly chosen amount
         in the range 20-120 ms.
         """
-        delay = millis_to_seconds(random.randint(*self._first_random_delay_interval))
-        next_time = now + delay
-        self._next_time = {check_type_: next_time for check_type_ in self._types}
-
-    def millis_to_wait(self, now: float_) -> float:
-        """Returns the number of milliseconds to wait for the next event."""
-        # Wait for the type has the smallest next time
-        next_time = min(self._next_time.values())
-        return 0 if next_time <= now else next_time - now
-
-    def reschedule_type(self, type_: str_, next_time: float_) -> bool:
-        """Reschedule the query for a type to happen sooner."""
-        if next_time >= self._next_time[type_]:
-            return False
-        self._next_time[type_] = next_time
-        return True
-
-    def _force_reschedule_type(self, type_: str_, next_time: float_) -> None:
-        """Force a reschedule of a type."""
-        self._next_time[type_] = next_time
-
-    def process_ready_types(self, now: float_) -> List[str]:
+        start_delay = millis_to_seconds(random.randint(*self._first_random_delay_interval))
+        self._loop = loop
+        loop.call_later(start_delay, self._process_ready_types)
+
+    def stop(self) -> None:
+        """Stop the scheduler."""
+        if self._next_run:
+            self._next_run.cancel()
+            self._next_run = None
+
+    def schedule(self, pointer: DNSPointer) -> None:
+        """Schedule a query for a pointer."""
+        expire_time = pointer.get_expiration_time(_EXPIRE_REFRESH_TIME_PERCENT)
+        scheduled_query = _ScheduledQuery(pointer.alias, pointer.name, expire_time)
+        self._next_scheduled_for_name[pointer.name] = scheduled_query
+        heappush(self._query_heap, scheduled_query)
+
+    def cancel(self, pointer: DNSPointer) -> None:
+        """Cancel a query for a pointer."""
+        self._next_scheduled_for_name[pointer.name].cancelled = True
+        del self._next_scheduled_for_name[pointer.name]
+
+    def reschedule(self, pointer: DNSPointer) -> None:
+        """Reschedule a query for a pointer."""
+        self.cancel(pointer)
+        self.schedule(pointer)
+
+    def _process_ready_types(self) -> None:
         """Generate a list of ready types that is due and schedule the next time."""
-        if self.millis_to_wait(now):
-            return []
+        if TYPE_CHECKING:
+            assert self._loop is not None
+        now_millis = current_time_millis()
+
+        if self._startup_queries_sent < STARTUP_QUERIES:
```

Comment:
This is an inefficient design. We should switch to calling a new function once we finish the startup queries instead of branching here